### ISSUE TITLE
The 'out'-parameter in torch.matmul() works for 'cuda' device but not for 'cpu'

### ISSUE THREAD

**Issue by drnil** [pytorch/pytorch#114350]
### 🐛 Describe the bug

The call to torch.matmul() below produces an incorrect zero result when using the 'out' keyword and a 'cpu' device. However, it works correctly on a 'cuda' device.

    import torch
    torch.autograd.set_grad_enabled(False)

    def test(dev):
        w = torch.tensor([[1, 1]],
                         device=torch.device(dev),
                         dtype=torch.float)
        g = torch.tensor([[1, 0],
                          [0, 1]],
                          device=torch.device(dev),
                          dtype=torch.float)
        w = torch.matmul(w,g)
        print(dev+':', w)
        w[:] = torch.matmul(w,g)
        print(dev+':', w)
        # The following call produces an incorrect result for dev='cpu':
        torch.matmul(w,g,out=w)
        print(dev+':', w)

    test('cuda')
    test('cpu')

Outputs:

    cuda: tensor([[1., 1.]], device='cuda:0')
    cuda: tensor([[1., 1.]], device='cuda:0')
    cuda: tensor([[1., 1.]], device='cuda:0')
    cpu: tensor([[1., 1.]])
    cpu: tensor([[1., 1.]])
    cpu: tensor([[0., 0.]])

### Versions

Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Pro
GCC version: (x86_64-posix-seh, Built by strawberryperl.com project) 8.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.6 (tags/v3.10.6:9c7b4bd, Aug  1 2022, 21:53:49) [MSC v.1932 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19045-SP0
Is CUDA available: N/A
CUDA runtime version: 10.1.105
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2070
Nvidia driver version: 537.42
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

CPU:
Architecture=9
CurrentClockSpeed=3600
DeviceID=CPU0
Family=207
L2CacheSize=2048
L2CacheSpeed=
Manufacturer=GenuineIntel
MaxClockSpeed=3600
Name=Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
ProcessorType=3
Revision=

Versions of relevant libraries:
[pip3] No relevant packages
[conda] Could not collect

**Comment by larsr** [comment]
I can confirm this.  I recreated the this bug on Google Colab.

```
import torch

a = torch.tensor([[1.,1.]])
b = torch.tensor([[1.,0.],[0.,1.]])
torch.matmul(a,b, out=a)
print(a)
# tensor([[0., 0.]])   # this is wrong

a = torch.tensor([[1.,1.]])
b = torch.tensor([[1.,0.],[0.,1.]])
a, b = a.to('cuda'), b.to('cuda')  # doing it on the GPU instead
torch.matmul(a,b, out=a)
print(a)
# tensor([[1., 1.]], ...)   # this is right
```

**Comment by colesbury** [comment]
Is this a new regression in 2.1.0? Or did we have this previously?

**Comment by drnil** [comment]
Apologies for not writing it in my original report, but my torch version is 2.1.1 .

**Comment by larsr** [comment]
I get the same error in version 2.0.0

**Comment by malfet** [comment]
I think in-place matmul just does not make much sense, we probably should check for it and return an error
Fun fact that following can return correct/incorrect results based on MKL version/CPU you are running on
```
% python3 -c "import torch;x=torch.arange(9, dtype=torch.float32, device='cpu').reshape(3,3);print(torch.mm(x, x, out=x))"
```

**Comment by malfet** [comment]
Probably related https://github.com/pytorch/pytorch/issues/87551#issuecomment-1288179570 (from @ptrblck )

**Comment by larsr** [comment]
> I think in-place matmul just does not make much sense

Sorry, how do you mean? In-place operations are generally useful for memory effiency reasons. Is it something related to matrix multiplication in particular?

**Comment by malfet** [comment]
> Sorry, how do you mean? In-place operations are generally useful for memory effiency reasons. Is it something related to matrix multiplication in particular?

I'm not an expert on the subject, but are there an in-place GEMM algorithm that wouldn't need a O(m*n) workspace?

As far as I remember for cpu implementation PyTorch relies on a BLASS gemm algorithm, that does not guarantee to work in place this producing incorrect results.

**Comment by drnil** [comment]
One common case where in-place matmul is important is when you want to update a large matrix by multiplication with a small matrix, such as when performing a convolution.

**Comment by malfet** [comment]
@drnil Correct, and convolution algorithms are likely to work in place, but torch.mm != torch.nn.conv2d

**Comment by drnil** [comment]
@malfet I wasn't thinking of the very specific NN use of convolution but more generally. There are many, many such applications. For example, computing the effect of a quantum gate on the state of a quantum computer.

**Comment by larsr** [comment]
I looked at Conv2d but it doesn't seem to have an option to work "inplace".

But either way, the annoying thing with the bug is that inplace (seems to) work on GPU but not on CPU.  But if it does not make sense on CPU then it probably also doesn't make sense on GPU either, or am I misunderstanding?

What would be the best way to handle this?  Is it a bug or is it something that is unsupported?  In my opinion, it comes down to the philosophy of the API.  Two ways to handle this  are

1. keep the `out=` argument but implement it in a slower but correct version behind the scenes? (philosophy: expose a high level api that lets one think on the level of linear algebra) or
2. Don't falsely imply that `inplace` multiplication is possible, and thus remove the argument if it is indeed buggy, (philosophy: exposing the limits of the low level implementation to let/force the user develop workarounds)

**Comment by drnil** [comment]
@larsr @malfet A quick check in networld indicates that a common method to implement this kind of in-place multiplication is to use repeated level-2 BLAS vector-matrix in-place multiplication. This should use only little additional memory (O(m)?).

Another clever trick is to use LU decomposition of the typically small, not in-place matrix (G in my example and B in larsr's example), and then use two level-3 BLAS calls for in-place multiplication (TRMM?) by triangular matrices, i.e. WG = (WL)U. For W input and output to be the same size, G has to be square and can therefore always be LU decomposed. If G is rectangular so that W shrinks, G could perhaps be zero padded to square format.

Anyway, if it works on a GPU, I cannot see why it shouldn't work on a CPU. But since GPUs are typically much more memory constrained than CPUs, the really important thing is that it works on the GPU, so please don't remove that! It can be worked around on the CPU (but with additional memory consumption), as illustrated in my example.

Btw, a sincere thank you to all of you who work on developing and maintaining PyTorch! You have made my life *so* much easier and more creative!

**Comment by malfet** [comment]
@ptrblck will check how cuBLAS handles in-place op, but we should probably error out
At the very least we should mention in the document that behavior is undefined
